hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/SafeModeRuleFactory.java

```diff
@@ -0,0 +1,131 @@
+
+
+  private static final Logger LOG = LoggerFactory.getLogger(SafeModeRuleFactory.class);
+
+  // TODO: Move the rule names to respective rules. (HDDS-11798)
+  private static final String CONT_EXIT_RULE = "ContainerSafeModeRule";
+  private static final String DN_EXIT_RULE = "DataNodeSafeModeRule";
+  private static final String HEALTHY_PIPELINE_EXIT_RULE =
+      "HealthyPipelineSafeModeRule";
+  private static final String ATLEAST_ONE_DATANODE_REPORTED_PIPELINE_EXIT_RULE =
+      "AtleastOneDatanodeReportedRule";
+
+  private final ConfigurationSource config;
+  private final SCMContext scmContext;
+  private final EventQueue eventQueue;
+
+  // TODO: Remove dependency on safeModeManager (HDDS-11797)
+  private final SCMSafeModeManager safeModeManager;
+  private final PipelineManager pipelineManager;
+  private final ContainerManager containerManager;
+
+  private final List<SafeModeExitRule<?>> safeModeRules;
+  private final List<SafeModeExitRule<?>> preCheckRules;
+
+  private static SafeModeRuleFactory instance;
+
+  private SafeModeRuleFactory(final ConfigurationSource config,
+                              final SCMContext scmContext,
+                              final EventQueue eventQueue,
+                              final SCMSafeModeManager safeModeManager,
+                              final PipelineManager pipelineManager,
+                              final ContainerManager containerManager) {
+    this.config = config;
+    this.scmContext = scmContext;
+    this.eventQueue = eventQueue;
+    this.safeModeManager = safeModeManager;
+    this.pipelineManager = pipelineManager;
+    this.containerManager = containerManager;
+    this.safeModeRules = new ArrayList<>();
+    this.preCheckRules = new ArrayList<>();
+    loadRules();
+  }
+
+  private void loadRules() {
+    // TODO: Use annotation to load the rules. (HDDS-11730)
+    safeModeRules.add(new ContainerSafeModeRule(CONT_EXIT_RULE, eventQueue, config,
+        containerManager, safeModeManager));
+    SafeModeExitRule<?> dnRule = new DataNodeSafeModeRule(DN_EXIT_RULE, eventQueue, config, safeModeManager);
+    safeModeRules.add(dnRule);
+    preCheckRules.add(dnRule);
+
+    // TODO: Move isRuleEnabled check to the Rule implementation. (HDDS-11799)
+    if (config.getBoolean(
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK,
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT)
+        && pipelineManager != null) {
+
+      safeModeRules.add(new HealthyPipelineSafeModeRule(HEALTHY_PIPELINE_EXIT_RULE,
+          eventQueue, pipelineManager, safeModeManager, config, scmContext));
+      safeModeRules.add(new OneReplicaPipelineSafeModeRule(
+          ATLEAST_ONE_DATANODE_REPORTED_PIPELINE_EXIT_RULE, eventQueue,
+          pipelineManager, safeModeManager, config));
+    }
+
+  }
+
+  public static SafeModeRuleFactory getInstance() {
+    if (instance != null) {
+      return instance;
+    }
```

Comment:
For thread-safety,
- `getInstance()` should be `synchronized`, OR
- `instance` should be `volatile`, and assigned to local variable first

Reply:
Same suggestion.  

Except this, others looks good to me.

Reply:
Not sure how I missed adding `synchronized` here. Thanks for the catch @adoroszlai  & @ChenSammi.